src/transform/src/lib.rs

```diff
@@ -86,47 +87,74 @@ macro_rules! any {
 /// manipulated by the transforms.
 #[derive(Debug)]
 pub struct TransformCtx<'a> {
+    /// The global ID for this query (if it exists).
+    pub global_id: Option<GlobalId>,
     /// The indexes accessible.
     pub indexes: &'a dyn IndexOracle,
     /// Statistical estimates.
     pub stats: &'a dyn StatisticsOracle,
-    /// The global ID for this query (if it exists).
-    pub global_id: Option<&'a GlobalId>,
-    /// Whether or not to eagerly apply delta joins.
-    pub enable_eager_delta_joins: bool,
+    /// Features passed to the enclosing `Optimizer`.
+    pub features: &'a OptimizerFeatures,
+    /// Typechecking context.
+    pub typecheck_ctx: &'a SharedContext,
     /// Transforms can use this field to communicate information outside the result plans.
-    pub dataflow_metainfo: &'a mut DataflowMetainfo,
+    pub df_meta: &'a mut DataflowMetainfo,
 }
 
 impl<'a> TransformCtx<'a> {
-    /// Generates a `TransformArgs` instance for the given `IndexOracle` with no `GlobalId`
-    pub fn dummy(dataflow_metainfo: &'a mut DataflowMetainfo) -> Self {
+    /// Generates a [`TransformCtx`] instance for the local MIR optimization
+    /// stage.
+    ///
+    /// Used to call [`Optimizer::optimize`] on a
+    /// [`Optimizer::logical_optimizer`] in order to transform a stand-alone
+    /// [`MirRelationExpr`].
+    pub fn local(
+        features: &'a OptimizerFeatures,
+        typecheck_ctx: &'a typecheck::SharedContext,
+        df_meta: &'a mut DataflowMetainfo,
+    ) -> Self {
         Self {
             indexes: &EmptyIndexOracle,
             stats: &EmptyStatisticsOracle,
             global_id: None,
-            enable_eager_delta_joins: false,
-            dataflow_metainfo,
+            features,
+            typecheck_ctx,
+            df_meta,
         }
     }
 
-    /// Generates a `TransformArgs` instance for the given `IndexOracle` and `StatisticsOracle` with
-    /// a `GlobalId`
-    pub fn with_config_and_metainfo(
+    /// Generates a [`TransformCtx`] instance for the global MIR optimization
+    /// stage.
+    ///
+    /// Used to call [`dataflow::optimize_dataflow`].
+    pub fn global(
```

Comment:
If you look at the sequence of commits there is an intermediate stage where actually things are as suggested by you, but then I am pulling the `TransformCtx::global(...)` call outside of `optimize_dataflow` into the `Optimize` implementations in `mz_adapter.

Mostly this is done for consistency between the local and the lobal pass, and because I don't want us to touch the signature of `optimize_dataflow` every time when the `TransformCtx` contents change—the `Optimize` implementations in `mz_adapter` have everything needed to construct a `TransformCtx`.

Reply:
Ok